### ISSUE TITLE
Store API: Attribute Terms return wrong parent

### ISSUE THREAD

**Issue by sunyatasattva** [woothemes/woocommerce#42471]
## Describe the bug

Attribute terms from the Store API have a `parent` key which is always `0`.

## To reproduce

0. Make sure your store has attributes with terms in them.
1. Make a Rest API request to `wc/store/v1/products/attributes/1/terms`.
2. Notice that the response includes a `parent` key which is always `0`.

## Expected behavior

The `parent` key should likely be the id of the parent attribute of that term (i.e. terms returned by `wc/store/v1/products/attributes/1/terms` should have `parent: 1`).

**Comment by albarin** [comment]
@sunyatasattva I've looked into this and I think `parent: 0` is correct. The "product attribute" (e.g. `Color` or `Size`) is actually a taxonomy, not a term itself. So the terms under it don't have a parent term, the same way as first-level terms under "Product categories" (e.g. `Clothing`) don't have a parent either (only second-level terms have, like `Accessories` or `Hoodies`).
Does it make sense? 🤔

**Comment by sunyatasattva** [comment]
@albarin thanks for checking! Maybe there is a misunderstanding:

The `wc/store/v1/products/attributes/1/terms` endpoint, in an example store, returns a list of terms under the `Color` taxonomy. Here is the response:

```
[
	{
		"id": 22,
		"name": "Blue",
		"slug": "blue",
		"description": "",
		"parent": 0,
		"count": 4
	},
	{
		"id": 28,
		"name": "Gray",
		"slug": "gray",
		"description": "",
		"parent": 0,
		"count": 3
	},
	{
		"id": 23,
		"name": "Green",
		"slug": "green",
		"description": "",
		"parent": 0,
		"count": 3
	},
	{
		"id": 24,
		"name": "Red",
		"slug": "red",
		"description": "",
		"parent": 0,
		"count": 4
	},
	{
		"id": 29,
		"name": "Yellow",
		"slug": "yellow",
		"description": "",
		"parent": 0,
		"count": 1
	}
]
```

So, yes, while the taxonomy “Color” (id: 1) should have `parent: 0`, these terms all have `parent: 0`, which seems incorrect to me since “Color” has `id: 1`. Or did I misunderstand your point?

**Comment by albarin** [comment]
I understand your point, that's what I thought as well. Let me try to do a better explanation of my thoughts with an example.

The `wp-json/wp/v2/product_cat` endpoint lists the terms under the `product_cat` taxonomy, and it returns:
<img width="608" alt="Screenshot 2023-02-27 at 10 56 02" src="https://user-images.githubusercontent.com/186112/221532347-e9ded866-b617-4b7a-83c5-e490a9142274.png">
You can see first-level terms like `Decor` or `Clothing` don't have a parent, only `Accessories` has one (because is a term under `Clothing`). 

In the same way, the `wc/store/v1/products/attributes/1/terms` endpoint lists the terms under the `pa_color` taxonomy, since they all are first-level terms their parent is 0 as well.
<img width="239" alt="Screenshot 2023-02-27 at 11 02 27" src="https://user-images.githubusercontent.com/186112/221533569-3d1c6af8-762d-4276-9c46-158c07c1209d.png">
`Color` and `Size` are not terms but taxonomies (`pa_color` and `pa_size`), term parents can only be other terms I think.

That could be why it's not documented in the [docs](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/src/StoreApi/docs/product-attribute-terms.md).
I've tried to find some issues related and I found this https://github.com/woocommerce/woocommerce/issues/12530#issuecomment-265420651

I'm not totally sure about all this either, I might be completely wrong 😬

**Comment by sunyatasattva** [comment]
Mm, I think I understand what you're saying, and I wonder if that's intended however. It seems to me that the `parent` of a term, should represent its parent `taxonomy` id.

Also considering that, AFAIK, while taxonomies can have nested taxonomies (such as the Clothing > Accessories example you gave), I don't think terms support hierarchy (which is what the comment you linked also seem to suggest). In this case it means that the `parent` key will in all cases be `0`, which makes it pretty useless.

We have this issue, for example, in the Attribute selector component in which we represent terms under their proper taxonomy, and we have to assign the parent manually.

**Comment by albarin** [comment]
>Also considering that, AFAIK, while taxonomies can have nested taxonomies (such as the Clothing > Accessories example you gave), I don't think terms support hierarchy (which is what the comment you linked also seem to suggest). In this case it means that the parent key will in all cases be 0, which makes it pretty useless.

I think terms in general support hierarchy (Clothing and Accessories are terms, not taxonomies). Only Product Attribute (which is a special taxonomy) terms don't support hierarchy.

>We have this issue, for example, in the Attribute selector component in which we represent terms under their proper taxonomy, and we have to assign the parent manually.

Could you point me to that example, I don't know where's the Attribute selector component that you are mentioning 😅

**Comment by sunyatasattva** [comment]
> I think terms in general support hierarchy (Clothing and Accessories are terms, not taxonomies). Only Product Attribute (which is a special taxonomy) terms don't support hierarchy.

Mm, ok I see what you're saying. Also, forgive me, I phrased it wrong: there are **hierarchical taxonomies**, whose _terms_ support _child terms_. I now understand that Attributes are taxonomies which return other taxonomies which have terms, and the [`parent`](https://github.com/WordPress/wordpress-develop/blob/7c965e781f297e9f47f3abd3835c613eb41c0765/src/wp-includes/class-wp-term.php#L82) key is defined as the ID of the parent's **term**, or 0.

So, you're right, the API is correct.

However, the API is not returning useful information in the form of the [`term_taxonomy_id`](https://github.com/WordPress/wordpress-develop/blob/7c965e781f297e9f47f3abd3835c613eb41c0765/src/wp-includes/class-wp-term.php#L58), which is what I thought the `parent` was. So, if I now understand Attributes correctly, Color is a `term` of taxonomy Attribute, and Blue is a `term` of taxonomy Color.

So, I suppose we could change this issue to a feature request or enhancement and not a bug, as the `parent` is correct, but I'd still think it's useful for the REST API to return also this extra info. What do you think?

> Could you point me to that example, I don't know where's the Attribute selector component that you are mentioning 😅

[Here is the component, adding the parent manually](https://vscode.dev/github/woocommerce/woocommerce-gutenberg-products-block/blob/15d1d7815385d0c93945aa7d25a0e9ec3f600c2b/assets/js/hocs/with-attributes.js#L91).

**Comment by albarin** [comment]
> So, I suppose we could change this issue to a feature request or enhancement and not a bug, as the parent is correct, but I'd still think it's useful for the REST API to return also this extra info. What do you think?

Yes, agree, it's clearly confusing, and having the extra info would be quite useful. Thank you for the discussion! 😄

**Comment by mikejolley** [comment]
All terms share the same schema for consistency; `TermSchema`. The description of which is: `Parent term ID, if applicable.`

If you think this description can be made clearer it should be changed there, but I don't think we should create different schema for different types of terms if possible.

> However, the API is not returning useful information in the form of the [term_taxonomy_id](https://github.com/WordPress/wordpress-develop/blob/7c965e781f297e9f47f3abd3835c613eb41c0765/src/wp-includes/class-wp-term.php#L58), which is what I thought the parent was.

We could introduce `term_taxonomy_id`, but what is the intended use case for that field? Anything you can lookup with `term_taxonomy_id` you can also lookup with the term ID.

**Comment by sunyatasattva** [comment]
> We could introduce term_taxonomy_id, but what is the intended use case for that field? Anything you can lookup with term_taxonomy_id you can also lookup with the term ID.

Hey thanks for chiming in @mikejolley! The use-case for terms being aware of their own parent is to build components like [the one linked above](https://github.com/woocommerce/woocommerce-blocks/blob/15d1d7815385d0c93945aa7d25a0e9ec3f600c2b/assets/js/hocs/with-attributes.js#L91). I'd say it'd be beneficial to somehow have that information returned by the API so that it is clear from an API response where does a term belong to.

**Comment by mikejolley** [comment]
@sunyatasattva it's not clear to me what that component is referring to. If its a parent term ID, that is is what we're providing already if the terms are hierarchical. Product Attributes do not fit this criteria because the taxonomies are defined as non hierarchical.

**Comment by sunyatasattva** [comment]
Sorry for not being clear. My use case was specifically for the `SearchListControl` component when it uses Product Attributes. I understand that Product Attributes are not hierarchical, but I'd like for each term returned from the API to be aware which taxonomy id is the parent (e.g. if I get “Blue”, I want a key in the API to tell me that it belongs to “Color”).

Does this make more sense?

**Comment by mikejolley** [comment]
@sunyatasattva some, however it looks like when you're making the request, the prerequisite is that you already know the taxonomy ID because its part of the request. I don't think we have an endpoint that could return multiple attributes across multiple taxonomies.

**Comment by sunyatasattva** [comment]
> however it looks like when you're making the request, the prerequisite is that you already know the taxonomy ID because its part of the request. 

Well, if you make a request to `store/products/:id` you are getting a response of a product back, but that product still has the `id` property, even if you knew it for the request. The same happens if you request: `store/products/attributes/:id`.

**Comment by samrat-biswas-git** [comment]
Reviewed and confirmed this is still valid. Marking as ready for engineering.

Current evidence: The issue's core claim was resolved in discussion: attribute terms are non-hierarchical terms, so `parent: 0` is correct under the shared term schema. Current trunk still documents `parent` as `Parent term ID, if applicable`, matching that interpretation.

**Comment by gigitux** [comment]
> [@sunyatasattva](https://github.com/sunyatasattva) some, however it looks like when you're making the request, the prerequisite is that you already know the taxonomy ID because its part of the request. I don't think we have an endpoint that could return multiple attributes across multiple taxonomies.

I agree with this comment. What’s the value of returning the ID here?

I see @sunyatasattva’s point, but given that this feature request originated from us and no one from the community has chimed in, I’d close this issue and mark it as “not planned.” We can always revisit it in the future if there’s a clear need for it.
